### PR TITLE
Logging: add output to azure blob

### PR DESCRIPTION
**Pull request by Gerhut**
Depends on #976 to set the tag of logs, which stands for the blob name.

```
 spent 0:04:03.895252 in executing 29 cases ['test_inference_job.test_inference_job_running', 'test_distributed_jo
b.test_blobfuse', 'test_distributed_job.test_distributed_job_env', 'test_distributed_job.test_distributed_job_ssh', 'test_distributed_job.test_distributed_non_pree
mptable_job_running', 'test_distributed_job.test_distributed_preemptable_job_running', 'test_distributed_job.test_distributed_with_default_cmd', 'test_distributed_
job.test_fault_tolerance', 'test_distributed_job.test_job_directory', 'test_distributed_job.test_ssh_cuda_visible_devices', 'test_regular_job.test_batch_kill_jobs'
, 'test_regular_job.test_batch_op_jobs', 'test_regular_job.test_blobfuse', 'test_regular_job.test_data_job_running', 'test_regular_job.test_do_not_expose_private_k
ey', 'test_regular_job.test_fault_tolerance', 'test_regular_job.test_job_fail', 'test_regular_job.test_list_all_jobs', 'test_regular_job.test_op_job', 'test_regula
r_job.test_regular_job_custom_ssh_key', 'test_regular_job.test_regular_job_env', 'test_regular_job.test_regular_job_ssh', 'test_regular_job.test_regular_non_preemp
table_job_running', 'test_regular_job.test_regular_preemptable_job_running', 'test_regular_job.test_ssh_cpu_job_cuda_visible_devices', 'test_regular_job.test_ssh_d
o_not_expose_private_key', 'test_regular_job.test_ssh_multi_gpu_job_cuda_visible_devices', 'test_regular_job.test_ssh_one_gpu_job_cuda_visible_devices', 'test_regu
lar_job.test_sudo_installed'], 0 failed
```

### CODE REVIEW

File: src/azure-blob-adapter/wsgi.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    :param PlainRequest request:
+    '''
+    try:
+        blob_name = request.headers['X-Tag']
+    except KeyError:
+        message = 'X-Tag header is required.'
+        logger.error(message)
+        return Response(message, status=400)
+
+    while True:
```
Comment (xudifsd) [inline src/azure-blob-adapter/wsgi.py]: I think using `while True` in production is very dangerous. If we want to retry, we'd better specify a retry limit, liket 3. Even a very large retry count is better than infinity.

File: src/docker-images/azure-blob-adapter/prebuild.sh [hunk 2]
```diff
@@ -0,0 +1 @@
+cp -a ../../../../azure-blob-adapter/* ./
```
Comment (xudifsd) [inline src/docker-images/azure-blob-adapter/prebuild.sh]: Why not just put content in docker-images?

File: src/utils/JobLogUtils.py [hunk 3]
```diff
@@ -15,45 +16,112 @@ def TryParseCursor(cursor):
...
+            if len(documents) > 0:
+                cursor = '.'.join(str(i) for i in documents[-1]["sort"])
+            else:
+                cursor = None
+
+            return (pod_logs, cursor)
+        except Exception:
+            logger.exception("Request elasticsearch failed")
+            return ({}, None)
+else:
```
Comment (xudifsd) [inline src/utils/JobLogUtils.py]: Maybe add a logging in each branch to show which backend is chosen to make debug faster?

File: src/ClusterBootstrap/template/RestfulAPI/config.yaml [hunk 4]
```diff
@@ -99,4 +99,14 @@ default_memoryrequest: {{cnf["default_memoryrequest"]}}
...
\ No newline at end of file
+{% endif %}
+
+{% if cnf["azureBlobLog"]["enabled"] %}
+logging: azureBlob
+{% elif cnf["elasticsearch"]["enabled"] %}
+logging: elasticsearch
+{% endif %}
+
+azureBlobLog:
```
[the remark this answers is not in the document]
Comment (Gerhut) [reply src/ClusterBootstrap/template/RestfulAPI/config.yaml]: Actually I saw almost all styles of field names in config, have we the team already decided a target name style for all to migrate? If so, I will start to migrate all of them.

File: src/ClusterBootstrap/hello-azure-blob.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+from json import loads
+from azure.storage.blob import AppendBlobService
+
+connection_string = 'DefaultEndpointsProtocol=https;AccountName=dltsdevlogs;AccountKey=/Yx5H4M3rTvIhhWtsTztR+119s4972rsAVG8u4UHiJsFc8XGWXHrlFxbwMwUu0E5/vU23c3/C+SrgeMcUEYN4w==;EndpointSuffix=core.windows.net'
```
[the remark this answers is not in the document]
Comment (Gerhut) [reply src/ClusterBootstrap/hello-azure-blob.py]: By mistake, removed and changed password

File: src/azure-blob-adapter/wsgi.py (hunk 1 above)
Comment (Gerhut) [reply src/azure-blob-adapter/wsgi.py]: Clarified and removed while-loop

File: src/utils/JobLogUtils.py (hunk 3 above)
Comment (Gerhut) [reply src/utils/JobLogUtils.py]: Added

File: src/docker-images/azure-blob-adapter/prebuild.sh (hunk 2 above)
Comment (Gerhut) [reply src/docker-images/azure-blob-adapter/prebuild.sh]: I see, just copy from what other services do

Comment (Gerhut) [reply src/docker-images/azure-blob-adapter/prebuild.sh]: Can I do soft link for the directory?

Comment (Gerhut) [reply src/docker-images/azure-blob-adapter/prebuild.sh]: Moved

File: src/ClusterBootstrap/template/RestfulAPI/config.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (xudifsd) [reply src/ClusterBootstrap/template/RestfulAPI/config.yaml]: I prefer snake style. @Anbang-Hu ?

File: src/utils/JobLogUtils.py (hunk 3 above)
Comment (xudifsd) [reply src/utils/JobLogUtils.py]: No, I meant add log outside of function, so the log only shown once. e.g.

```
if config.get("logging") == 'azureBlob':
    logger.info('Requesting Azure Blob to get log of the job %s', jobId)
    from azure.storage.blob import AppendBlobService
...
```

File: src/ClusterBootstrap/template/RestfulAPI/config.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Gerhut) [reply src/ClusterBootstrap/template/RestfulAPI/config.yaml]: Changed

File: src/utils/JobLogUtils.py (hunk 3 above)
Comment (Gerhut) [reply src/utils/JobLogUtils.py]: Got it

Comment (Gerhut) [reply src/utils/JobLogUtils.py]: I see

**Comment by xudifsd** [comment]
@Anbang-Hu please review when you're free

File: src/ClusterManager/joblog_manager.py [hunk 6]
```diff
@@ -37,51 +34,42 @@ def create_log(logdir='/var/log/dlworkspace'):
...
 
-        for (container_id, log_text) in container_logs.items():
+        for (pod_name, log_text) in pod_logs.items():
             try:
-                containerLogPath = os.path.join(
-                    jobLogDir, "log-conatainer-" + container_id + ".txt")
-                with open(containerLogPath, 'a') as f:
+                podLogPath = os.path.join(
+                    jobLogDir, "log-pod-" + pod_name + ".txt")
+                with open(podLogPath, 'a') as f:
```
Comment (Anbang-Hu) [inline src/ClusterManager/joblog_manager.py]: Can we add `encoding="utf-8"` for safety?

File: src/ClusterManager/joblog_manager.py [hunk 7]
```diff
@@ -37,51 +34,42 @@ def create_log(logdir='/var/log/dlworkspace'):
...
                     f.write(log_text)
-                os.system("chown -R %s %s" % (userId, containerLogPath))
+                os.system("chown -R %s %s" % (userId, podLogPath))
             except Exception:
-                logger.exception("write container log failed")
+                logger.exception("write pod log failed")
```
Comment (Anbang-Hu) [inline src/ClusterManager/joblog_manager.py]: Can we add `jobId` in the exception message as an indication to which job failed as well?

File: src/ClusterManager/joblog_manager.py (hunk 6 above)
Comment (Gerhut) [reply src/ClusterManager/joblog_manager.py:68]: Will rebase

Comment (Gerhut) [reply src/ClusterManager/joblog_manager.py]: Rebased

File: src/ClusterManager/joblog_manager.py (hunk 7 above)
Comment (Gerhut) [reply src/ClusterManager/joblog_manager.py:72]: Added

**Pull request merged**